### DISCUSSION TITLE
How to work with import-maps

### DISCUSSION THREAD

**Discussion by jerrychan7** [creativelifeform/three-nebula#168] [Q&A]
Although the script tag can be used normally, I want the same import style with the es module, and I lost the key step of importing three-nebula as an es module. For the sake of simplicity and ease of learning, I did not use any module bundler, such as Webpack and Browserify, but used the [import-maps feature](https://github.com/WICG/import-maps) that is experimenting. This feature works well with Three.js, but I tried a lot and still cannot work with three-nebula.

I use npm to install three.js and three-nebula:
```bash
npm i --save three three-nebula
```
and then, I map the import path:
```html
<script type="importmap">
    {
        "imports": {
            "three": "./node_modules/three/build/three.module.js",
            "three/": "./node_modules/three/",
            "three-nebula": "./node_modules/three-nebula/build/three-nebula.js"
        }
    }
</script>
```
and usage:
```js
import * as THREE from "three";
import * as Nebula from "three-nebula";
```
above will get an error `Uncaught SyntaxError: The requested module 'three-nebula' does not provide an export named 'default'`.

I also tried to map the path to `./node_modules/three-nebula/build/esm/index.js`, but got 404 errors due to the import mechanism of the browser environment. The node.js or babel environment allows the `.js` file extension to be omitted, and the file name is allowed to be omitted when there is an `index.js` file in the directory. For example, `import './core/index.js'` can be written as `import './core/index'` or `import './core'`. These omissions make the import-maps configuration invalid. The module name in the browser is a URL and must start with `/`, `./`, or `../`, or be parsable as an absolute URL, to identify a URL.

---

***Or*** is there any way to use es-style import while keeping it simple without use import-maps?

---

P.S. My mother tongue is not English. I have expressed my meaning as much as possible. Please let me know if you have any questions or things you can't understand. Thank you!

**Comment by rohan-deshpande** [comment]
Can you post the code you've used to try and use `Nebula`? It might be `Nebula.Nebula` or `Nebula.System` I can't recall right now. A https://codesandbox.io example for your issue would be appreciated so I can help debug

**Reply by jerrychan7** [reply]
Of course, I can show my code.

Copy felow in HTML file:
```html
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <title>three-nebula test</title>
    <style> body { margin: 0; } </style>
    <script type="importmap">
        {
            "imports": {
                "three": "./node_modules/three/build/three.module.js",
                "three/": "./node_modules/three/",
                "three-nebula": "./node_modules/three-nebula/build/three-nebula.js"
            }
        }
    </script>
    <!-- Modifying the above path to `./node_modules/three-nebula/build/esm/index.js` will also get errors. -->
</head>
<body>
    <script type="module">
        import * as THREE from "three";
        import * as Nebula from "three-nebula";

        // below will show a green box
        let renderer = new THREE.WebGLRenderer();
        renderer.setSize(window.innerWidth, window.innerHeight);
        document.body.appendChild(renderer.domElement);
        let scene = new THREE.Scene();
        scene.add(new THREE.Mesh(new THREE.BoxGeometry(), new THREE.MeshBasicMaterial({ color: 0x00ff00, })));
        renderer.render(scene, new THREE.PerspectiveCamera(75, window.innerWidth / window.innerHeight, 0.1, 1000).translateZ(5));
    </script>
</body>
</html>
```

and use npm to install three.js and three-nebula in the same folder with the above HTML file:
```bash
npm i --save three three-nebula
```

Run a server according to your environment in the same folder to avoid cross-origin requests error. For example:
* python: `python -m http.server 3000`
* node.js:
    ```bash
    npm i http-serve -g
    http-serve -p 3000
    ```
